datumaro/components/project.py

```diff
@@ -2,486 +2,2399 @@
-
-    def put(self, item, id=None, subset=None, path=None):
-        if path is None:
-            path = item.path
-
-        if path:
-            source = path[0]
-            # TODO: reverse remapping
-            self._sources[source].put(item, id=id, subset=subset, path=path[1:])
-
-        if id is None:
-            id = item.id
-        if subset is None:
-            subset = item.subset
-
-        item = item.wrap(path=path)
-        if subset not in self._subsets:
-            self._subsets[subset] = self.Subset(self, subset)
-        self._subsets[subset].items[id] = item
-        self._length = None
-
-        return item
-
-    def save(self, save_dir=None, merge=False, recursive=True,
-            save_images=False):
-        if save_dir is None:
-            assert self.config.project_dir
-            save_dir = self.config.project_dir
-            project = self._project
+    try:
+        module = import_module.__func__()
+    except ImportError:
+        module = None
+
+    def _git_dir(self):
+        return osp.join(self._project_dir, '.git')
+
+    def __init__(self, project_dir, repo=None):
+        self._project_dir = project_dir
+        self.repo = repo
+
+        if repo is None and \
+                osp.isdir(project_dir) and osp.isdir(self._git_dir()):
+            self.repo = self.module.Repo(project_dir)
+
+    @property
+    def initialized(self):
+        return self.repo is not None
+
+    def init(self):
+        if self.initialized:
+            return
+
+        repo = self.module.Repo.init(path=self._project_dir)
+        repo.config_writer() \
+            .set_value("user", "name", "User") \
+            .set_value("user", "email", "<>") \
+            .release()
+        # gitpython does not support init, use git directly
+        repo.git.init()
+
+        self.repo = repo
+
+    @property
+    def refs(self) -> List[str]:
+        return [t.name for t in self.repo.refs]
+
+    @property
+    def tags(self) -> List[str]:
+        return [t.name for t in self.repo.tags]
+
+    def push(self, remote=None):
+        args = [remote] if remote else []
+        remote = self.repo.remote(*args)
+        branch = self.repo.head.ref.name
+        if not self.repo.head.ref.tracking_branch():
+            self.repo.git.push('--set-upstream', remote, branch)
         else:
-            merge = True
+            remote.push(branch)
+
+    def pull(self, remote=None):
+        args = [remote] if remote else []
+        return self.repo.remote(*args).pull()
+
+    def check_updates(self, remote=None) -> List[str]:
+        args = [remote] if remote else []
+        remote = self.repo.remote(*args)
+        prev_refs = {r.name: r.commit.hexsha for r in remote.refs}
+        remote.update()
+        new_refs = {r.name: r.commit.hexsha for r in remote.refs}
+        updated_refs = [(prev_refs.get(n), new_refs.get(n))
+            for n, _ in (set(prev_refs.items()) ^ set(new_refs.items()))]
```

Comment:
Doesn't this algorithm mean that a ref name can occur twice in `update_refs` if it's present in both `prev_refs` and `new_refs` with different commit hashes?

Reply:
Maybe. It doesn't really matter now, because remotes are not supported yet. I don't want to drop these functions, because they already took lots of efforts to find out the approximate implementations. Maybe, it would be better to put not implemented errors into the unused ones.

Reply:
> Maybe, it would be better to put not implemented errors into the unused ones.

It would certainly be helpful for figuring out what functionality isn't used yet.

Reply:
Done